### ISSUE TITLE
The rules no-used-vars and no-use-before-defined show an error where non should be reported. 

### ISSUE THREAD

**Issue by burka** [eslint/eslint#9118]
The rules no-used-vars and no-use-before-defined show an error where non should be reported. 

I attached a Test-Case for no-used-vars with #9117.

If you rename the isHidden function or remove the return value it works as expected.

**What is the purpose of this pull request? (put an "X" next to item)**
[x] Bug Test-Case without fix
**Tell us about your environment**

* **ESLint Version:**
4.4.1

* **Node Version:**
6.11.1 and 6.11.2

* **npm Version:**
3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
  "parserOptions": {
    "sourceType": "module",
    "ecmaVersion": 6,
    "ecmaFeatures": {
      "modules": true,
      "arrowFunctions": true,
      "blockBindings": true,
      "destructuring": true,
      "classes": true
    }
  },
  "extends": [
    "eslint:recommended"
  ],
  "env": {
    "browser": true,
    "es6": true
  },
  "plugins": [
  ],
  "rules": {
    "no-use-before-define": "error",
    "no-unused-vars": "error"
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```
function isHidden(something) {
    return !something;
}

function doSomething(map) {
    const isHidden = isHidden(map['something']);
}

doSomething();
```

**What did you expect to happen?**
I expected just a warning that isHidden is unused.

**What actually happened? Please include the actual, raw output from ESLint.**
I got an error that isHidden is used before defined and unused.
```
  1:10  error  'isHidden' is defined but never used       no-unused-vars
  6:22  error  'isHidden' was used before it was defined  no-use-before-define

✖ 2 problems (2 errors, 0 warnings)

```
**What changes did you make? (Give an overview)**
I added a Test-Case that shows the error but **does not fix it**.

**Is there anything you'd like reviewers to focus on?**
Maybe this should only be committed together with a fix?

**Comment by burka** [comment]
Haha, is it due to the fact that `const isHidden` has the same name aus the `function isHidden` and thus overrides the function definition before being defined?

**Comment by not-an-aardvark** [comment]
Thanks for the pull request!

> Haha, is it due to the fact that `const isHidden` has the same name as the `function isHidden` and thus overrides the function definition before being defined?

Yes, I think so. There are two `isHidden` variables in your code: one from the function declaration, and one from the variable declaration. The variable from the variable declaration shadows the variable from the function declaration, so all references to `isHidden` inside the `doSomething` block are referring to that variable. This means that (a) the `isHidden` variable from the function declaration is never used, and (b) the first line of the `doSomething` block uses the variable `isHidden` from the variable declaration as part of the initial assignment to that variable, so it's used before it's defined. So I think this is working as intended.

**Comment by burka** [comment]
Yes, you are right :+1:
